linkerd/app/outbound/src/ingress.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{http, stack_labels, tcp, trace_labels, Config};
+use linkerd2_app_core::{
+    config::{ProxyConfig, ServerConfig},
+    discovery_rejected, drain, errors, http_request_l5d_override_dst_addr, metrics,
+    opencensus::proto::trace::v1 as oc,
+    profiles,
+    spans::SpanConverter,
+    svc::{self},
+    transport::{self, io, listen, tls},
+    Addr, AddrMatch, Error, TraceContext,
+};
+use tokio::sync::mpsc;
+use tracing::info_span;
+
+/// Routes HTTP requests according to the l5d-dst-override header.
+///
+/// Forwards TCP connections without discovery/routing (or mTLS).
+///
+/// This is only intended for Ingress configurations, where we assume all
+/// outbound traffic is either
+pub fn stack<P, T, TSvc, H, HSvc, I>(
+    config: &Config,
+    profiles: P,
+    tcp: T,
+    http: H,
+    metrics: &metrics::Proxy,
+    span_sink: Option<mpsc::Sender<oc::Span>>,
+    drain: drain::Watch,
+) -> impl svc::NewService<
+    listen::Addrs,
+    Service = impl tower::Service<
+        I,
+        Response = (),
+        Error = impl Into<Error>,
+        Future = impl Send + 'static,
+    > + Send
+                  + 'static,
+> + Send
+       + 'static
+where
+    I: io::AsyncRead + io::AsyncWrite + io::PeerAddr + std::fmt::Debug + Unpin + Send + 'static,
+    T: svc::NewService<tcp::Endpoint, Service = TSvc> + Unpin + Clone + Send + Sync + 'static,
+    TSvc: tower::Service<
+            io::PrefixedIo<transport::metrics::SensorIo<I>>,
+            Response = (),
+            Error = Error,
+        > + Clone
+        + Send
+        + 'static,
+    TSvc::Future: Send,
+    H: svc::NewService<http::Logical, Service = HSvc> + Unpin + Send + Clone + 'static,
+    HSvc: tower::Service<
+            http::Request<http::boxed::Payload>,
+            Response = http::Response<http::boxed::Payload>,
+            Error = Error,
+        > + Send
+        + 'static,
+    HSvc::Future: Send,
+    P: profiles::GetProfile<Addr> + Unpin + Clone + Send + 'static,
+    P::Future: Unpin + Send,
+    P::Error: Send,
+{
+    let Config {
+        allow_discovery,
+        proxy:
+            ProxyConfig {
+                server: ServerConfig { h2_settings, .. },
+                dispatch_timeout,
+                max_in_flight_requests,
+                detect_protocol_timeout,
+                buffer_capacity,
+                cache_max_idle_age,
+                ..
+            },
+    } = config.clone();
+
+    let http = svc::stack(http)
```

Comment:
I haven't looked closely yet but it feels like there's _probably_ some common code between the ingress-mode and normal mode http stacks that we might want to share. nbd if not, but adding more places that future changes will have to update always worries me a little bit...

Reply:
I agree, but deduping things is easier said than done. If you can get it working, I'll merge it ;)

Reply:
i'm not sure if it makes sense here or not, but i can take a look.